### PR TITLE
Build & Deploy Window Fixes

### PR DESCRIPTION
**Pull request by StephenHodgson**
- Fixes https://github.com/Microsoft/HoloToolkit-Unity/issues/688 Building only for D3D. Now uses build settings.
- Fixes https://github.com/Microsoft/HoloToolkit-Unity/issues/685,  https://github.com/Microsoft/HoloToolkit-Unity/issues/612, & https://github.com/Microsoft/HoloToolkit-Unity/issues/615 Added Support for MSBuild 15+ while keeping backwards compatibility.
- Fixes https://github.com/Microsoft/HoloToolkit-Unity/issues/601 Fixed Build Failure due to App Name with Space.
- Fixes https://github.com/Microsoft/HoloToolkit-Unity/issues/583 Added UWP SDK pop up so users can choose sdk version in build window.
- Fixes https://github.com/Microsoft/HoloToolkit-Unity/issues/255 Fixed issue with bad builds returning error code 0.
- Fixes https://github.com/Microsoft/HoloToolkit-Unity/issues/189 Added ability to select a certificate and input password (Users can only create new certs from the Player Settings Window).
- Added workaround to unity bug where `project.json` and `project.lock.json` are not copied correctly to the target build path. (See [UwpBuildTest](https://github.com/StephenHodgson/UwpBuildTest))
- Added Scrollbar for APPX package list.
- Added checks for SLN existing before attempting to make APPX.
- Disabled buttons if HoloLens is not detected via usb.
- Merged changes from [RS2_MixedReality](https://github.com/Microsoft/HoloToolkit-Unity/pull/572/files#diff-915d8ae44ef16ebac775ab81e0cdf317) branch that specifically addressed changes to Build Window.
- Changed Build Configuration from a string field to an enum popup.
- Added a button to open the appx packages location.

image

### CODE REVIEW

File: Assets/HoloToolkit/Build/Editor/CopyDirectoryInfo.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+//
+
+namespace HoloToolkit.Unity
+{
+    public class CopyDirectoryInfo
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Build/Editor/CopyDirectoryInfo.cs]: Feel like this belongs in Utilities...

File: Assets/HoloToolkit/Utilities/Scripts/Editor/ExternalProcess.cs [hunk 2]
```diff
@@ -31,10 +31,10 @@ public class ExternalProcess : IDisposable
 
         private IntPtr mHandle;
 
-        /*
-        * First some static utility functions, used by some other code as well.
-        * They are related to "external processes" so they appear here.
-        */
+        /// <summary>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Editor/ExternalProcess.cs]: This class probably belongs in Utilities too.

File: Assets/HoloToolkit/Build/Editor/BuildSLNUtilities.cs [hunk 3]
```diff
@@ -288,7 +288,16 @@ public static void PerformBuild_CommandLine()
             var buildInfo = new BuildInfo
             {
                 // Use scenes from the editor build settings.
-                Scenes = EditorBuildSettings.scenes.Where(scene => scene.enabled).Select(scene => scene.path)
+                Scenes = EditorBuildSettings.scenes.Where(scene => scene.enabled).Select(scene => scene.path),
+
+                // Configure a post build action that will compile the generated solution
```
Comment (jwittner) [inline Assets/HoloToolkit/Build/Editor/BuildSLNUtilities.cs]: Nit: Not sure this comment is accurate.

File: Assets/HoloToolkit/Build/Editor/BuildSLNUtilities.cs [hunk 4]
```diff
@@ -288,7 +288,16 @@ public static void PerformBuild_CommandLine()
...
+                PostBuildAction = (innerBuildInfo, buildError) =>
+                {
+                    if (string.IsNullOrEmpty(buildError))
+                    {
+                        EditorApplication.Exit(1);
```
Comment (jwittner) [inline Assets/HoloToolkit/Build/Editor/BuildSLNUtilities.cs]: Yay!

Comment (StephenHodgson) [inline Assets/HoloToolkit/Build/Editor/BuildSLNUtilities.cs]: Hey shouldn't this be` !string.IsNullOrEmpty(buildError)`?

**Comment by StephenHodgson** [comment]
Okay It's ready

**Comment by keveleigh** [comment]
It looks like you've merged some changes from the RS2 branch without bringing them in via the original (mostly massive) commits. How will we know when all the changes are in, if we do it this way?

**Comment by StephenHodgson** [comment]
There won't be conflicts or changes.

**Comment by StephenHodgson** [comment]
If anyone has suggestions on how to make the Build process run on a new thread, I'm all ears.

**Comment by StephenHodgson** [comment]
I ran into an issue where Build SLN, Build APPX, and then Install button does not trigger an install.

Watch for an update!

**Comment by StephenHodgson** [comment]
Could I get some reviews for this?  I was hoping to have it in before making the next release.

**Comment by StephenHodgson** [comment]
@keveleigh, we can either patch, or cherry pick the commits in either direction. I tried doing this already without much success. Ultimately the real way to know that the change are in is if there aren't any merge conflicts and if there are, checking if there are differences.  Hopefully by the end, they'll be happy.

**Comment by StephenHodgson** [comment]
@keveleigh, I was able to include the history for the RS2 branch.

File: Assets/HoloToolkit/Build/Editor/BuildSLNUtilities.cs [hunk 5]
```diff
@@ -268,14 +255,17 @@ public static void ParseBuildCommandLine(ref BuildInfo buildInfo)
 
                     buildInfo.WSASdk = (WSASDK)Enum.Parse(typeof(WSASDK), wsaSdkArg);
                 }
+                else if (string.Equals(arguments[i], "-usaUwpSdk", StringComparison.InvariantCultureIgnoreCase))
```
Comment (Strepto) [inline Assets/HoloToolkit/Build/Editor/BuildSLNUtilities.cs]: Should not this be "-wsaUwpSdk" ?

**Pull request merged**